### PR TITLE
Fix invulnerable end crystals: MC-108513

### LINKED ISSUE #5712: Enderdragon Respawning Glitching
### Expected behavior

The ender dragon respawns and heals from the end crystals, then this phase ends and you can fight the dragon while blowing up all the end crystals as well.

### Observed/Actual behavior

When you use pistons to move the endcrystals that are healing the dragon as it is respawning, the ender dragon respawn phase glitches and is forever in the healing animation and stays like this. Additionally, the healing end crystals are invincible and players are using pistons to move them to the overworld where they cannot be destroyed by a player in survival mode.

### Steps/models to reproduce

Respawn the ender dragon by hand placing the necessary crystals and then while it is in the healing animation use pistons to move an end crystal with beam.

### Plugin list

ProtocolLib, ViaVersion, ViaBackwards

### Paper version

This server is running Paper version git-Paper-655 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)

### Agreements

- [X] I am running the latest version of Paper available from https://papermc.io/downloads.
- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] My version of Minecraft is supported by Paper.

### Other

https://bugs.mojang.com/browse/MC-108513

This appears to be a bug that was supposedly patched in 1.13 but people are reporting it is happening again in 1.16.x and the 1.17 snapshots

image

### PR DESCRIPTION
**Pull request by Machine-Maker**
[MC-108513](https://bugs.mojang.com/browse/MC-108513)

Fixes #5712 (I think)

### CODE REVIEW

**Comment by Machine-Maker** [comment]
Makes #5733 irrelevant as well

File: Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/WorldGenEnder.java b/src/main/java/net/minecraft/world/level/levelgen/feature/WorldGenEnder.java
+index dd623702131eaa1a65937a19a0e986e865322258..1960fc029262d54c6a5ee05c99d86428386cba4a 100644
+--- a/src/main/java/net/minecraft/world/level/levelgen/feature/WorldGenEnder.java
++++ b/src/main/java/net/minecraft/world/level/levelgen/feature/WorldGenEnder.java
+@@ -42,7 +42,7 @@ public class WorldGenEnder extends WorldGenerator<WorldGenFeatureEndSpikeConfigu
+         return (List) WorldGenEnder.a.getUnchecked(i);
+     }
+ 
+-    public boolean a(GeneratorAccessSeed generatoraccessseed, ChunkGenerator chunkgenerator, Random random, BlockPosition blockposition, WorldGenFeatureEndSpikeConfiguration worldgenfeatureendspikeconfiguration) {
++    public boolean generate(GeneratorAccessSeed generatoraccessseed, ChunkGenerator chunkgenerator, Random random, BlockPosition blockposition, WorldGenFeatureEndSpikeConfiguration worldgenfeatureendspikeconfiguration) { // Paper - decompile fix
```
Comment (Proximyst) [inline Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch]: What was fixed?

File: Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                 entityendercrystal.setInvulnerable(false);
+                 entityendercrystal.setBeamTarget((BlockPosition) null);
+             }
++            // Paper start
++            } else {
++                worldgenender_spike.crystal.setInvulnerable(false);
++                worldgenender_spike.crystal.setBeamTarget(null);
++                worldgenender_spike.crystal = null;
++            }
++            // Paper end
```
Comment (Proximyst) [inline Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch]: Why is this state even possible? I'm no fan of such a solution.

File: Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch (hunk 1 above)
Comment (Machine-Maker) [reply Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch]: The name of the method.

File: Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch (hunk 2 above)
Comment (Machine-Maker) [reply Spigot-Server-Patches/0751-Fix-invulnerable-end-crystals.patch]: I don’t know if this state is possible. I’m unclear on what happens if the chunks are unloaded during a dragon respawn, or any other fringe behavior so I thought best to leave the old logic in. 

Is your “no fan of such a solution” directed at the whole idea of the spikes keeping a reference to the entity, or at leaving the old reset logic?

**Review by Proximyst — APPROVED** [review]
Looks fine

**Pull request merged**